### ISSUE TITLE
Range indexing fails

### ISSUE THREAD

**Issue by iamed2** [invenia/Intervals.jl#78]
```
julia> using Dates, TimeZones, Intervals

julia> st = HE(ZonedDateTime(2014,1,1, tz"America/Winnipeg"))
AnchoredInterval{-1 hour,ZonedDateTime}(ZonedDateTime(2014, 1, 1, tz"America/Winnipeg"), Inclusivity(false, true))

julia> r = st:Hour(1):(st + Day(1))
AnchoredInterval{-1 hour,ZonedDateTime}(ZonedDateTime(2014, 1, 1, tz"America/Winnipeg"), Inclusivity(false, true)):1 hour:AnchoredInterval{-1 hour,ZonedDateTime}(ZonedDateTime(2014, 1, 2, tz"America/Winnipeg"), Inclusivity(false, true))

julia> collect(r)[1:2]
2-element Array{AnchoredInterval{-1 hour,ZonedDateTime},1}:
 AnchoredInterval{-1 hour,ZonedDateTime}(ZonedDateTime(2014, 1, 1, tz"America/Winnipeg"), Inclusivity(false, true))
 AnchoredInterval{-1 hour,ZonedDateTime}(ZonedDateTime(2014, 1, 1, 1, tz"America/Winnipeg"), Inclusivity(false, true))

julia> r[1:2]
Error showing value of type StepRangeLen{AnchoredInterval{-1 hour,ZonedDateTime},AnchoredInterval{-1 hour,ZonedDateTime},Hour}:
ERROR: MethodError: Cannot `convert` an object of type Hour to an object of type ZonedDateTime
Closest candidates are:
  convert(::Type{T}, ::Interval{T}) where T at /Users/ericdavies/.julia/packages/Intervals/52k6l/src/interval.jl:124
  convert(::Type{T}, ::AnchoredInterval{P,T}) where {P, T} at /Users/ericdavies/.julia/packages/Intervals/52k6l/src/anchoredinterval.jl:174
  convert(::Type{T}, ::T) where T at essentials.jl:167
  ...
Stacktrace:
 [1] AnchoredInterval{-1 hour,ZonedDateTime}(::Hour) at /Users/ericdavies/.julia/packages/Intervals/52k6l/src/anchoredinterval.jl:79
 [2] step(::StepRangeLen{AnchoredInterval{-1 hour,ZonedDateTime},AnchoredInterval{-1 hour,ZonedDateTime},Hour}) at ./range.jl:501
 [3] show(::IOContext{REPL.Terminals.TTYTerminal}, ::StepRangeLen{AnchoredInterval{-1 hour,ZonedDateTime},AnchoredInterval{-1 hour,ZonedDateTime},Hour}) at ./range.jl:712
 [4] show(::IOContext{REPL.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::StepRangeLen{AnchoredInterval{-1 hour,ZonedDateTime},AnchoredInterval{-1 hour,ZonedDateTime},Hour}) at ./show.jl:7
 [5] display(::REPL.REPLDisplay{REPL.LineEditREPL}, ::MIME{Symbol("text/plain")}, ::StepRangeLen{AnchoredInterval{-1 hour,ZonedDateTime},AnchoredInterval{-1 hour,ZonedDateTime},Hour}) at /Users/ericdavies/.julia/packages/OhMyREPL/GFHgr/src/output_prompt_overwrite.jl:6
 [6] display(::REPL.REPLDisplay, ::Any) at /Users/ericdavies/repos/julia1p2/usr/share/julia/stdlib/v1.2/REPL/src/REPL.jl:136
 [7] display(::Any) at ./multimedia.jl:323
```

Range indexing appears to be taking a `StepRange` and making a `StepRangeLen`, which we don't handle effectively.

I think this might be a bug in `step(::StepRangeLen)`.

**Comment by iamed2** [comment]
I think that bug is there, but we can also avoid it by never making a `StepRangeLen`. We can prevent that by defining the traits `OrderStyle` and `ArithmeticStyle` to be the same as `AbstractTime`. We should also add `RangeStepStyle` while we're at it.

**Comment by iamed2** [comment]
See https://github.com/JuliaLang/julia/blob/master/base/range.jl and https://github.com/JuliaLang/julia/blob/master/base/traits.jl

**Comment by omus** [comment]
As of [Julia 1.5.0-DEV.207](https://github.com/JuliaLang/julia/pull/34563) (7d92a3aaed) this issue appears fixed:

```julia 
julia> r[1:2]
AnchoredInterval{Hour(-1),ZonedDateTime}(ZonedDateTime(2014, 1, 1, tz"America/Winnipeg"), Inclusivity(false, true)):Hour(1):AnchoredInterval{Hour(-1),ZonedDateTime}(ZonedDateTime(2014, 1, 1, 1, tz"America/Winnipeg"), Inclusivity(false, true))
```
Do we want to try and fix this on earlier versions of Julia?

**Comment by iamed2** [comment]
We should figure out the correct traits. If that fixes it for older versions of Julia, that would be a bonus.

**Comment by omus** [comment]
Another example of this issue but on Intervals 1.3 you just see a deprecation warning:
```julia
julia> using Intervals

julia> x = AnchoredInterval{-1}(1):1:AnchoredInterval{-1}(5)
AnchoredInterval{-1,Int64,Open,Closed}(1):1:AnchoredInterval{-1,Int64,Open,Closed}(5)

julia> y = x[1:4];

julia> y
┌ Warning: `convert(T, interval::AnchoredInterval{P,T})` is deprecated for intervals which are not closed with coinciding endpoints. Use `anchor(interval)` instead.
│   caller = AnchoredInterval at anchoredinterval.jl:95 [inlined]
└ @ Core ~/.julia/dev/Intervals/src/anchoredinterval.jl:95
┌ Warning: `convert(T, interval::AnchoredInterval{P,T})` is deprecated for intervals which are not closed with coinciding endpoints. Use `anchor(interval)` instead.
│   caller = AnchoredInterval at anchoredinterval.jl:95 [inlined]
└ @ Core ~/.julia/dev/Intervals/src/anchoredinterval.jl:95
AnchoredInterval{-1,Int64,Open,Closed}(1):AnchoredInterval{-1,Int64,Open,Closed}(1):AnchoredInterval{-1,Int64,Open,Closed}(4)
```

**Comment by omus** [comment]
I attempted to use traits to fix this problem but that didn't actually seem to be the source of the failure. I instead implemented the fix in #132 and made a [new issue for implementing traits](#133).
